mountpoint-s3-client/src/s3_crt_client.rs

```diff
@@ -89,12 +93,30 @@ impl S3CrtClient {
 
         let mut client_bootstrap = ClientBootstrap::new(&allocator, &bootstrap_options).unwrap();
 
-        let creds_options = CredentialsProviderChainDefaultOptions {
-            bootstrap: &mut client_bootstrap,
-        };
+        let mut client_config = ClientConfig::new();
 
-        let mut creds_provider = CredentialsProvider::new_chain_default(&allocator, &creds_options).unwrap();
-        let signing_config = init_default_signing_config(region, &mut creds_provider);
+        if !config.no_sign_request {
+            let credentials_provider = match config.profile_name_override {
+                Some(profile_name_override) => {
+                    let credentials_profile_options = CredentialsProviderProfileOptions {
+                        bootstrap: &mut client_bootstrap,
+                        profile_name_override: Some(&profile_name_override),
+                    };
+                    CredentialsProvider::new_profile(&allocator, credentials_profile_options)
+                        .map_err(|_| CredentialsError::ProfileNotFound)
+                }
+                None => {
+                    let credentials_chain_default_options = CredentialsProviderChainDefaultOptions {
+                        bootstrap: &mut client_bootstrap,
```

Comment:
(existing code, so it may be a wider question, but you also used this on line 102)
Why the exclusive ref here? And why client_bootstrap needs to be mut in the first place?
Shouldn't we either:
1. pass by shared ref and rely on the crt to acquire what it needs to hold on to when we call `aws_credentials_provider_new_chain_default` inside `new_chain_default`?
2. or take ownership and clone here, which is what we are doing e.g. on line 129?

Reply:
Two parts to this:
1. We pass by reference instead of cloning for two reasons:
  a. to avoid the refcount increment/decrement overhead (very premature optimization)
  b. to avoid baking the fact that these things are cheaply cloneable into our API, in case we wanted to change that for some reason in the future.
3. Given that we're passing by reference, we pass with `&mut` because the CRT API expects a `*mut` pointer. Technically we could cast the `&` reference into a `*mut` pointer, and it would be fine here, but it's sketchy in general.

I think probably what we should do is reverse our choice in (1) and making all this stuff pass by value with a ref count increment when cloning. But we can do that in a separate PR.